### PR TITLE
Limit VSO power of attorney users’ ability to view decision issues - testing covers prod outage

### PR DESCRIPTION
**Pull request by eileen-nava**
Resolves jira ticket [Show Only Request Issues for POA users until Decision Date ](https://vajira.max.gov/browse/CASEFLOW-1522).

Relates to PRs:
- [Limit VSO users' ability to view decision issues #16433](https://github.com/department-of-veterans-affairs/caseflow/pull/16433 )
- [Seed data for feature to restrict poa visibility #16466]( https://github.com/department-of-veterans-affairs/caseflow/pull/16466 )
- [Revert "Limit VSO users' ability to view decision issues" #16469](https://github.com/department-of-veterans-affairs/caseflow/pull/16469)
- [Revert "seed data for feature to restrict poa visibility" #16468]( https://github.com/department-of-veterans-affairs/caseflow/pull/16468)
[Sentry link with outage information](https://sentry.prod.appeals.va.gov/department-of-veterans-affairs/caseflow/issues/18929/).

### Description

Previous work for Jira Ticket 1522 caused a prod outage because changes in the appeal serializer broke the case details page.  The method `current_user`, which I added to the appeal serializer [after having the following comments exchange](https://github.com/department-of-veterans-affairs/caseflow/pull/16466#discussion_r666964587) and consulting the [Caseflow wiki](https://github.com/department-of-veterans-affairs/caseflow/wiki/Adding-a-Feature-Flag-with-FeatureToggle#getting-a-featureflag-state),  cannot be called in the appeal serializer.  Our automated tests did not catch this bug because `StubbableUser` has a `current_user` method, which returned nil in tests, rather than throwing an error, as it would outside of tests.

I replicated the error by mocking the error that `current_user` would throw in prod.  I also included a comment to provide context for the new test set-up.

As I discussed with other Echo engineers, there’s a risk that a second prod outage could slip by our automated tests if another engineer mistakenly uses the  `current_user` method in a file where it is not available.  It would be good to expand our tests to catch such an error.  However, I don’t think that is in the scope of this PR.

### Acceptance Criteria
- [ ] There is testing for the scenario where the method `current_user` throws an error
- [ ] Acceptance criteria from original PR [here](https://github.com/department-of-veterans-affairs/caseflow/pull/16433 ) 

### Testing Plan
1. Please see the original testing plan [here](https://github.com/department-of-veterans-affairs/caseflow/pull/16433)  

### User Facing Changes
 - [ ] Please see original PR [here](https://github.com/department-of-veterans-affairs/caseflow/pull/16433)

### CODE REVIEW

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+describe WorkQueue::AppealSerializer, :all_dbs do
+  context "when an appeal has a decision issue with a decision date in the future" do
+    let(:appeal) { create(:appeal, :decision_issue_with_future_date) }
+    subject { described_class.new(appeal, params: { user: user }) }
+
+    context "when a VSO user views an appeal" do
+      let(:user) { create(:user, :vso_role) }
```
Comment (n1zyy) [inline spec/models/serializers/work_queue/appeal_serializer_spec.rb:9]: It looks like this is only used in line 6, in which case it might read a little better if this `let(:user...` was defined right before that line.

(This is probably front of mind for me because I just worked on a spec and kept moving stuff between blocks.)

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      context "when the restrict_poa_visibility feature toggle is on" do
+        before { FeatureToggle.enable!(:restrict_poa_visibility) }
+        after { FeatureToggle.enable!(:restrict_poa_visibility) }
```
Comment (n1zyy) [inline spec/models/serializers/work_queue/appeal_serializer_spec.rb:13]:
```suggestion
        after { FeatureToggle.disable!(:restrict_poa_visibility) }
```

(I think? Maybe we just don't need this?)

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        end
+        after { FeatureToggle.enable!(:restrict_poa_visibility) }
+
+        describe "decision_issues" do
+          it "does display decision issues with a decision date in the future" do
+            expect(subject.serializable_hash[:data][:attributes][:decision_issues].count).to eq(1)
+          end
+
+          it "accesses user information without breaking on the user passed into the policy" do
+            expect { subject.serializable_hash[:data][:attributes][:decision_issues] }.not_to raise_error
```
Comment (n1zyy) [inline spec/models/serializers/work_queue/appeal_serializer_spec.rb:37]: If this was going to raise an error, wouldn't line 33 have hit it first?

**Review by n1zyy — COMMENTED** [review]
Overall this looks really well-done. I left some small suggestions.

I haven't had the cycles to start a manual review yet but can pick that up in the morning.

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb (hunk 1 above)
Comment (jcq) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:9]: It makes sense to have it defined within the context blocks if there isn't a "default" sort of state for it (if you are testing one vs the other, for instance); otherwise yeah — a default could be defined before the subject and then redefined in just the place it changes

File: app/policies/appeal_decision_issues_policy.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      visible_issues = @appeal.decision_issues.select do |issue|
+        # VSO users should not be able to see decision issues until the issue decision date
+        Time.now.utc > issue.caseflow_decision_date
+      end
+      modified_issues = visible_issues.map do |issue|
+        # VSO users should not be able to see decision issues' descriptions, regardless of the issue decision date
+        issue.description = nil
+        issue
+      end
+      modified_issues
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply app/policies/appeal_decision_issues_policy.rb:30]: Changed it. 👍

File: db/seeds/decision_issues.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    private
+
+    def vet
+      @vet ||= create(
+        :veteran,
+        file_number: 42_424_242,
+        first_name: "Joe",
+        last_name: "Dead",
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply db/seeds/decision_issues.rb:18]: I'll change the last name to "Doe" since the veteran is not deceased.

File: db/seeds/decision_issues.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      )
+    end
+
+    def build_original_appeal(veteran: vet, docket_type: "evidence_submission")
+      create(:appeal, :decision_issue_with_future_date, veteran: veteran, docket_type: docket_type)
+    end
+
+    def setup_decision_issue_seeds
+      build_original_appeal(veteran: vet, docket_type: "evidence_submission")
+    end
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply db/seeds/decision_issues.rb:29]: Implemented 👍

File: db/seeds/tasks.rb [hunk 7]
```diff
@@ -627,9 +627,9 @@ def create_tasks_at_acting_judge
       JudgeTeam.for_judge(judge).add_user(acting_judge)
 
       create_appeal_at_judge_assignment(judge: acting_judge)
-      create_task_at_attorney_review(create(:appeal), judge, acting_judge)
+      create_task_at_attorney_review(create(:appeal), judge, attorney)
       create_task_at_attorney_review(create(:appeal), acting_judge, attorney)
-      create_task_at_judge_review(create(:appeal), judge, acting_judge)
+      create_task_at_judge_review(create(:appeal), judge, attorney)
```
[the remark this answers is not in the document]
Comment (eileen-nava) [reply db/seeds/tasks.rb:632]: Here's a link to my previous [conversation with JC about this change](https://github.com/department-of-veterans-affairs/caseflow/pull/16466/files#r666956918)

TLDR: I changed `acting_judge` to `attorney` because when I ran `bundle exec rake db:seed`, it broke on this line.  The error message said that `acting_judge` is required to be of the type Attorney.  According to JC, an `acting_judge` is an attorney who temporarily has judge powers.

I can try to revert this change.  I am concerned it may break again when when I run `bundle exec rake db:seed`.

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb (hunk 1 above)
Comment (eileen-nava) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:9]: Makes sense!

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb (hunk 2 above)
Comment (eileen-nava) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:13]: Fixed, TY!

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb (hunk 3 above)
Comment (eileen-nava) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:37]: Very true!  I wanted to have a separate it block for this because I want the testing related to the prod outage to be as explicit as possible.  Any further thoughts on this?

Comment (n1zyy) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:36]: 🤔 I don't feel strongly. If it were me, I think I would remove this, and rely on the `.and_throw("Error!")` above to accomplish this. But I'm all for prioritizing readability if you think it's clearer this way.

File: db/seeds/tasks.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (n1zyy) [reply db/seeds/tasks.rb:632]: I am like 25% sure that this is just because the database wasn't cleared before running it and maybe it's not idempotent? I'm slightly worried just because I don't know a whole lot about the data this is trying to set up, or what our change to it would do. Is `db:seed` expected to work without having wiped the DB first, I wonder?

File: spec/models/serializers/work_queue/appeal_serializer_spec.rb (hunk 3 above)
Comment (eileen-nava) [reply spec/models/serializers/work_queue/appeal_serializer_spec.rb:36]: I hear that.  I'm going to leave the separate block, because I am feeling cautious.

**Review by n1zyy — APPROVED** [review]
Manual test checks out, too! Feature toggle operates as intended, and only impacts VSOs. :shipit: 💯
